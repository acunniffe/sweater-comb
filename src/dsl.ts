import {
  ApiCheckDsl,
  EntityRule,
  Result,
  DocsLinkHelper,
  newDocsLinkHelper,
  runCheck,
} from "@useoptic/api-checks";
import {
  ConceptualLocation,
  OpenApiHeaderFact,
  OpenApiKind,
  OpenApiOperationFact,
  IChange,
  IFact,
  OpenApiFieldFact,
  jsonPointerHelper,
  ILocation,
  OpenAPIV3,
} from "@useoptic/openapi-utilities";
import { genericEntityRuleImpl } from "@useoptic/api-checks/build/sdk/generic-entity-rule-impl";
import { ShouldOrMust } from "@useoptic/api-checks/build/sdk/types";
import { OpenApiResponseFact } from "@useoptic/openapi-utilities/build/openapi3/implementations/openapi3/openapi-traverser";

type SnykStablity = "wip" | "experimental" | "beta" | "ga";
type DateString = string; // YYYY-mm-dd
type ResourceName = string;

export interface SynkApiCheckContext {
  // Vervet provides context about the change itself. Since
  // Optic is analyzing two OpenAPI spec files, we need to tell it
  // when the change is supposed to happen, and the resource/version info
  // determined by the file's location in a directory structure.
  changeDate: DateString; // when the change did (or would, if proposed) occur
  changeResource: ResourceName; // the spec resource being changed
  changeVersion: {
    // the spec version being changed
    date: DateString;
    stability: SnykStablity;
  };

  // Vervet provides a mapping that indicates the resource version deprecation.
  // It has this information because it processes the entire source tree of
  // spec files.
  resourceVersions: {
    [ResourceName: string]: {
      // changeResource used to match this
      [DateString: string]: {
        // changeVersion.date used to match this
        [SnykStablity: string]: {
          // changeVersion.stability matches this
          deprecatedBy: {
            // the spec version that deprecates this one (if any) or null
            date: DateString;
            stability: SnykStablity; // could be higher stability than changed!
          } | null;
        };
      };
    };
  };
}

export interface SnykEntityRule<T, A>
  extends EntityRule<T, ConceptualLocation, SynkApiCheckContext, A> {}

export class SnykApiCheckDsl implements ApiCheckDsl {
  private checks: Promise<Result>[] = [];

  constructor(
    private nextFacts: IFact<any>[],
    private changelog: IChange<any>[],
    private nextJsonLike: OpenAPIV3.Document,
    private providedContext: SynkApiCheckContext
  ) {}

  checkPromises() {
    return this.checks;
  }

  getContext(location: ILocation): ConceptualLocation & SynkApiCheckContext {
    return {
      ...location.conceptualLocation,
      ...this.providedContext,
    };
  }

  get operations() {
    return genericEntityRuleImpl<
      OpenApiOperationFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.OperationObject
    >(
      OpenApiKind.Operation,
      this.changelog,
      this.nextFacts,
      (opFact) => `${opFact.method.toUpperCase()} ${opFact.pathPattern}`,
      (location) => this.getContext(location),
      (...items) => this.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(this.nextJsonLike, pointer)
    );
  }

  get context() {
    const change: IChange<SnykApiCheckDsl> = {
      location: {
        conceptualLocation: { path: "This Specification", method: "" },
        jsonPath: "/",
        conceptualPath: [],
        kind: "API",
      },
    };

    const value: ShouldOrMust<
      (
        context: SynkApiCheckContext,
        docs: DocsLinkHelper
      ) => Promise<void> | void
    > = {
      must: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          true,
          () => handler(this.providedContext, docsHelper)
        );
      },
      should: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          false,
          () => handler(this.providedContext, docsHelper)
        );
      },
    };

    return value;
  }

  get responses() {
    const dsl = this;

    return {
      ...genericEntityRuleImpl<
        OpenApiResponseFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.ResponsesObject
      >(
        OpenApiKind.Response,
        dsl.changelog,
        dsl.nextFacts,
        (response) => `response ${response.statusCode}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
      headers: genericEntityRuleImpl<
        OpenApiHeaderFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.HeaderObject
      >(
        OpenApiKind.ResponseHeader,
        dsl.changelog,
        dsl.nextFacts,
        (header) => `response header ${header.name}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
    };
  }

  get checkApiContext(): ShouldOrMust<
    (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
  > {
    const contextChangedHandler: (must: boolean) => ContextChangedRule["must"] =
      (must: boolean) => {
        return (statement, handler) => {
          const docsHelper = newDocsLinkHelper();
          const syntheticChange = {
            added: this.providedContext,
            location: {
              jsonPath: "/",
              conceptualPath: [],
              conceptualLocation: {
                path: "Entire Resource",
                method: "",
              },
              kind: "ContextRule",
            },
          };
          return runCheck(
            syntheticChange,
            docsHelper,
            "api lifeycle: ",
            statement,
            must,
            () => handler(this.providedContext, docsHelper)
          );
        };
      };

    return {
      must: contextChangedHandler(true),
      should: contextChangedHandler(false),
    };
  }

  get bodyProperties(): SnykEntityRule<
    OpenApiFieldFact,
    OpenAPIV3.SchemaObject
  > {
    const dsl = this;
    return genericEntityRuleImpl<
      OpenApiFieldFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.SchemaObject
    >(
      OpenApiKind.Field,
      dsl.changelog,
      dsl.nextFacts,
      (field) => `field ${field.key}`,
      (location) => dsl.getContext(location),
      (...items) => dsl.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
    );
  }
}

type ContextChangedRule = ShouldOrMust<
  (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
>;
