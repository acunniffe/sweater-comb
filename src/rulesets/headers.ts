import { SnykApiCheckDsl } from '../dsl';
const { expect } = require('chai');

export const rules = {
  headerNameCase: ({ responses }: SnykApiCheckDsl) => {
    responses.headers.requirement.must('be kebab case', ({ name }) => {
      const kebabCase = /^[a-z]+(-[a-z]+)+$/g;
      expect(kebabCase.test(name)).to.be.true;
    });
  },
};
