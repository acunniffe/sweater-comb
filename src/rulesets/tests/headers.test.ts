import { rules } from "../headers";
import { createTestDslFixture } from "@useoptic/api-checks";
import { SnykApiCheckDsl, SynkApiCheckContext } from "../../dsl";

// todo: fix copy/paste
const { compare } = createTestDslFixture<SnykApiCheckDsl, SynkApiCheckContext>(
  (input) => {
    return new SnykApiCheckDsl(
      input.nextFacts,
      input.changelog,
      input.nextJsonLike,
      input.context
    );
  }
);

// todo: fix copy/paste
const emptyContext: SynkApiCheckContext = {
  changeDate: "2021-10-10",
  changeResource: "Example",
  changeVersion: {
    date: "2021-10-10",
    stability: "ga",
  },
  resourceVersions: {},
};

describe("headers", () => {
  // todo: fix copy/paste
  const baseOpenAPI = {
    openapi: "3.0.1",
    paths: {
      "/example": {
        get: {
          responses: {},
        },
      },
    },
    info: { version: "0.0.0", title: "Empty" },
  };

  describe("name", () => {
    it("passes if kebab case", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec.paths!["/example"]!.get!.responses = {
            "200": {
              description: "",
              headers: {
                "good-header": {
                  schema: { type: "string" },
                },
              },
            },
          };
          return spec;
        })
        .withRule(rules.headerNameCase, emptyContext);

      expect(result.results[0].passed).toBeTruthy();
      expect(result).toMatchSnapshot();
    });

    it("fails if not kebab case", async () => {
      const result = await compare(baseOpenAPI)
        .to((spec) => {
          spec.paths!["/example"]!.get!.responses = {
            "200": {
              description: "",
              headers: {
                badHeader: {
                  schema: { type: "string" },
                },
              },
            },
          };
          return spec;
        })
        .withRule(rules.headerNameCase, emptyContext);

      expect(result.results[0].passed).toBeFalsy();
      expect(result).toMatchSnapshot();
    });
  });
});
