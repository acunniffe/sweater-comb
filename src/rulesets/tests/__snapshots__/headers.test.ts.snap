// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`headers name fails if not kebab case 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "added": Object {
        "name": "badHeader",
        "schema": Object {
          "type": "string",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "header": "badHeader",
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "headers",
          "badHeader",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/headers/badHeader",
        "kind": "response-header",
      },
    },
    Object {
      "added": Object {
        "description": "",
        "statusCode": 200,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1example/get/responses/200",
        "kind": "response",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "description": "",
              "headers": Object {
                "badHeader": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
            },
          },
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "header": "badHeader",
              "statusCode": "200",
            },
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "headers",
            "badHeader",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/headers/badHeader",
          "kind": "response-header",
        },
        "value": Object {
          "name": "badHeader",
          "schema": Object {
            "type": "string",
          },
        },
      },
      "condition": "be kebab case",
      "docsLink": undefined,
      "error": "expected false to be true",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "requirement for response-header: response header badHeader",
    },
  ],
}
`;

exports[`headers name passes if kebab case 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {},
        },
      },
    },
  },
  "changelog": Array [
    Object {
      "added": Object {
        "name": "good-header",
        "schema": Object {
          "type": "string",
        },
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "header": "good-header",
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
          "headers",
          "good-header",
        ],
        "jsonPath": "/paths/~1example/get/responses/200/headers/good-header",
        "kind": "response-header",
      },
    },
    Object {
      "added": Object {
        "description": "",
        "statusCode": 200,
      },
      "location": Object {
        "conceptualLocation": Object {
          "inResponse": Object {
            "statusCode": "200",
          },
          "method": "get",
          "path": "/example",
        },
        "conceptualPath": Array [
          "operations",
          "/example",
          "get",
          "responses",
          "200",
        ],
        "jsonPath": "/paths/~1example/get/responses/200",
        "kind": "response",
      },
    },
  ],
  "next": Object {
    "info": Object {
      "title": "Empty",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/example": Object {
        "get": Object {
          "responses": Object {
            "200": Object {
              "description": "",
              "headers": Object {
                "good-header": Object {
                  "schema": Object {
                    "type": "string",
                  },
                },
              },
            },
          },
        },
      },
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "inResponse": Object {
              "header": "good-header",
              "statusCode": "200",
            },
            "method": "get",
            "path": "/example",
          },
          "conceptualPath": Array [
            "operations",
            "/example",
            "get",
            "responses",
            "200",
            "headers",
            "good-header",
          ],
          "jsonPath": "/paths/~1example/get/responses/200/headers/good-header",
          "kind": "response-header",
        },
        "value": Object {
          "name": "good-header",
          "schema": Object {
            "type": "string",
          },
        },
      },
      "condition": "be kebab case",
      "docsLink": undefined,
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "requirement for response-header: response header good-header",
    },
  ],
}
`;
